quickwit/quickwit-index-management/src/index.rs

```diff
@@ -193,6 +196,46 @@ impl IndexService {
         Ok(deleted_splits)
     }
 
+    /// Deletes the indexes specified with `index_id_patterns`.
+    /// This is a wrapper of delete_index, and support index delete with index pattern
+    ///
+    /// * `index_id_patterns` - The target index Id patterns.
+    /// * `dry_run` - Should this only return a list of affected files without performing deletion.
+    pub async fn delete_indexes(
+        &self,
+        index_id_patterns: Vec<String>,
+        dry_run: bool,
+    ) -> Result<Vec<SplitInfo>, IndexServiceError> {
+        // get all index_ids by index_id_patterns
+        let indexes_metadata = self
+            .metastore()
+            .list_indexes_metadatas(ListIndexesQuery::IndexIdPatterns(index_id_patterns.clone()))
+            .await?;
+
+        if indexes_metadata.is_empty() {
+            return Err(IndexServiceError::Metastore(MetastoreError::NotFound(
+                EntityKind::Indexes {
+                    index_ids: index_id_patterns,
+                },
+            )));
+        }
+        let index_ids = indexes_metadata
+            .iter()
+            .map(|index_metadata| index_metadata.index_id())
+            .collect_vec();
+        info!(index_ids = ?index_ids);
+
+        // setup delete index tasks
+        let mut delete_index_tasks = Vec::new();
+        for index_id in index_ids {
+            delete_index_tasks.push(async move {
+                info!("delete_index:{}", index_id);
+                self.delete_index(index_id, dry_run).await
+            })
+        }
+        let delete_index_responses: Vec<Vec<SplitInfo>> = try_join_all(delete_index_tasks).await?;
```

Comment:
I'm a bit worried by this. It will return as soon as an error happens in one of the tasks. I think we should return the info that some indexes were deleted and some not. 

Thus, we should probably gather the split info for each index, and if there are some errors, return an error.

I think we should ditch the `Vec<SplitInfo>` and just keep tracks of successful deleted index + errors.

Reply:
Yes, try_join_all will return as soon as an error happens.
I agree we should return all of the successful deleted index. The tricky is what error should we return if more than one delete_index operation fails?

Reply:
This is what I have in mind:

```rust
     type IndexId = String;
      struct DeleteIndexesResponse {
          successes: Vec<IndexId>,
          failures: HashMap<IndexId, anyhow::Error>,
      }
      let delete_index_tasks= Vec::new();
      let indexes_metadata: Vec<IndexMetadata> = Vec::new(); 
      let successes: Vec<IndexId> = Vec::new();
      let failures: HashMap<IndexId, anyhow::Error> = Vec::new();
      for index_metadata in indexes_metadata {
          delete_index_tasks.push(async move { Ok::<_, anyhow::Error>("myindexid".to_string()) });
      }
      let mut stream = futures::stream::iter(delete_index_tasks).buffer_unordered(100);
      while let Some((index_id, delete_res)) = stream.next().await {
          match delete_res {
              Ok(_) => successes.push(index_id),
              Err(error) => {
                  failures.insert(index_id, error);
              }
          };
      }
      Ok(DeleteIndexesResponse {
          successes,
          failures,
      })
```

Reply:
@fmassot Thanks for the sample code
That's what I am doing right now.
I mean the tricky is what should we response in the rest handler and keep all of the other parts unchanged.(test cases, command line using the rest api)
I am thinking using the same return siguratues as delete_index, but using MetastoreError::Internal to include more detailed error information
Any suggestions?
```rust
 let mut delete_index_tasks = Vec::new();
        for index_id in index_ids {
            delete_index_tasks.push(async move {
                Ok::<_, _>(index_id, self.delete_index(index_id, dry_run).await)
            })
        }
        let mut delete_responses: HashMap<String, Vec<SplitInfo>> = HashMap::new();
        let mut delete_errors: HashMap<String, IndexServiceError> = HahsMap::new();
        let mut stream = futures::stream::iter(delete_index_tasks).buffer_unordered(100);
        while let Some((index_id, delete_response)) = stream.next().await {
            match delete_res {
                Ok(split_infos) => {
                    delete_responses.insert(index_id, split_infos)
                }
                Err(error) => {
                    delete_errors.insert(index_id, error)
                }
            }
        }

        if delete_errors.is_empty() {
            let mut concatenated_split_infos = Vec::new();
            for (_, split_info_vec) in delete_responses.into_iter() {
                concatenated_split_infos.extend(split_info_vec);
            }
            Ok(concatenated_split_infos)
        } else{
            return Err(IndexServiceError::Metastore(MetastoreError::Internal {
                message: format!("errors occurred when deleting indexes: %?", index_id_patterns),
                cause: format!("errors: {:?}\ndeleted indexes: {:?}", delete_errors, delete_responses)
            }));
        }
        ```

Reply:
I'm ok with this approach.